### ISSUE TITLE
Bevy v0.15.0 build fails with optimization option

### ISSUE THREAD

**Issue by amam3214** [bevyengine/bevy#16971]
## Bevy version
Bevy version = v0.15.0

## System information
A docker container on a windows machine.

## What was attempted
I was trying to build Bevy with the following Cargo.toml, recommended by the [set-up](https://bevyengine.org/learn/quick-start/getting-started/setup/#cargo-workspaces).
```toml
[package]
name = 'my_bevy_game'
version = '0.1.0'
edition = '2021'

[dependencies]
bevy = '0.15.0'

# Enable a small amount of optimization in the dev profile.
[profile.dev]
opt-level = 1

# Enable a large amount of optimization in the dev profile for dependencies.
[profile.dev.package."*"]
opt-level = 3
```

Testing this was done by building the following Dockerfile:
```Dockerfile
FROM rust:1.83-bullseye

RUN cargo new /home/bevy-v0_15_10
WORKDIR /home/bevy-v0_15_10

# Create Cargo.toml
RUN echo "[package] \n\
name = 'my_bevy_game' \n\
version = '0.1.0' \n\
edition = '2021' \n\
\n\
[dependencies] \n\
bevy = '0.15.0' \n\
 \n\
# Enable a small amount of optimization in the dev profile. \n\
[profile.dev] \n\
opt-level = 1 \n\
 \n\
# Enable a large amount of optimization in the dev profile for dependencies. \n\
[profile.dev.package.'*'] \n\
opt-level = 3 " > "Cargo.toml"

RUN apt-get update && apt-get install -y \
        libwayland-dev libxkbcommon-dev \
        g++ pkg-config libx11-dev libasound2-dev libudev-dev libxkbcommon-x11-0
```

and then running in my terminal:
```bash
docker build . -t bevy:v0.15.0
docker run -it bevy:v0.15.0
# now inside the docker container
cargo build
```
After almost compiling everything the following error is raised and the build fails and stops:
```
error: could not compile `bevy_render` (lib)

Caused by:
  process didn't exit successfully: `/usr/local/rustup/toolchains/1.83.0-x86_64-unknown-linux-gnu/bin/rustc --crate-name bevy_render --edition=2021 /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/bevy_render-0.15.0/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=156 --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no -C debuginfo=2 '--warn=clippy::unwrap_or_default' --warn=unused_qualifications --warn=unsafe_op_in_unsafe_fn --deny=unsafe_code --warn=unexpected_cfgs '--warn=clippy::undocumented_unsafe_blocks' '--allow=clippy::type_complexity' '--allow=clippy::too_long_first_doc_paragraph' '--warn=clippy::std_instead_of_core' '--warn=clippy::std_instead_of_alloc' '--warn=clippy::semicolon_if_nothing_returned' '--warn=clippy::ref_as_ptr' '--warn=clippy::redundant_else' '--warn=clippy::redundant_closure_for_method_calls' '--warn=clippy::ptr_cast_constness' '--warn=clippy::ptr_as_ptr' --warn=missing_docs '--warn=clippy::match_same_arms' '--warn=clippy::manual_let_else' '--warn=clippy::doc_markdown' '--warn=clippy::alloc_instead_of_core' --check-cfg 'cfg(docsrs_dep)' -C debug-assertions=on --cfg 'feature="hdr"' --cfg 'feature="ktx2"' --cfg 'feature="multi_threaded"' --cfg 'feature="webgl"' --check-cfg 'cfg(docsrs)' --check-cfg 'cfg(feature, values("basis-universal", "ci_limits", "dds", "detailed_trace", "exr", "hdr", "ios_simulator", "ktx2", "multi_threaded", "profiling", "shader_format_glsl", "shader_format_spirv", "spirv_shader_passthrough", "trace", "tracing-tracy", "webgl", "webgpu"))' -C metadata=c018e806ddc7dfc5 -C extra-filename=-c018e806ddc7dfc5 --out-dir /home/bevy-v0_15_10/target/debug/deps -L dependency=/home/bevy-v0_15_10/target/debug/deps --extern async_channel=/home/bevy-v0_15_10/target/debug/deps/libasync_channel-5010e979771a44f9.rmeta --extern bevy_app=/home/bevy-v0_15_10/target/debug/deps/libbevy_app-c41f08641878c131.rmeta --extern bevy_asset=/home/bevy-v0_15_10/target/debug/deps/libbevy_asset-db70c7fbbd812e7f.rmeta --extern bevy_color=/home/bevy-v0_15_10/target/debug/deps/libbevy_color-b63f193e85a44ec9.rmeta --extern bevy_core=/home/bevy-v0_15_10/target/debug/deps/libbevy_core-11adb1459002c6d9.rmeta --extern bevy_derive=/home/bevy-v0_15_10/target/debug/deps/libbevy_derive-a36bff48cbb6d803.so --extern bevy_diagnostic=/home/bevy-v0_15_10/target/debug/deps/libbevy_diagnostic-b1bc130631262eb8.rmeta --extern bevy_ecs=/home/bevy-v0_15_10/target/debug/deps/libbevy_ecs-761c0185a1c52c72.rmeta --extern bevy_encase_derive=/home/bevy-v0_15_10/target/debug/deps/libbevy_encase_derive-602b5776c49bc9d7.so --extern bevy_hierarchy=/home/bevy-v0_15_10/target/debug/deps/libbevy_hierarchy-46a7d4cc6e097bda.rmeta --extern bevy_image=/home/bevy-v0_15_10/target/debug/deps/libbevy_image-ca2a43002d3b51e0.rmeta --extern bevy_math=/home/bevy-v0_15_10/target/debug/deps/libbevy_math-b3da14759206059a.rmeta --extern bevy_mesh=/home/bevy-v0_15_10/target/debug/deps/libbevy_mesh-791d53d681ae28f7.rmeta --extern bevy_reflect=/home/bevy-v0_15_10/target/debug/deps/libbevy_reflect-febda5a40b921977.rmeta --extern bevy_render_macros=/home/bevy-v0_15_10/target/debug/deps/libbevy_render_macros-cb3e0b47b13bf923.so --extern bevy_tasks=/home/bevy-v0_15_10/target/debug/deps/libbevy_tasks-b8e4bf47ac43e925.rmeta --extern bevy_time=/home/bevy-v0_15_10/target/debug/deps/libbevy_time-84ba3b4de9446ff8.rmeta --extern bevy_transform=/home/bevy-v0_15_10/target/debug/deps/libbevy_transform-18e0bc892927c0af.rmeta --extern bevy_utils=/home/bevy-v0_15_10/target/debug/deps/libbevy_utils-a3c50cecef0864d2.rmeta --extern bevy_window=/home/bevy-v0_15_10/target/debug/deps/libbevy_window-7d2b86834f475524.rmeta --extern bytemuck=/home/bevy-v0_15_10/target/debug/deps/libbytemuck-9ef1f0f72b61bec0.rmeta --extern codespan_reporting=/home/bevy-v0_15_10/target/debug/deps/libcodespan_reporting-e3145b6b39ce778e.rmeta --extern derive_more=/home/bevy-v0_15_10/target/debug/deps/libderive_more-15fce4e52511f117.rmeta --extern downcast_rs=/home/bevy-v0_15_10/target/debug/deps/libdowncast_rs-cee32500dbd84e64.rmeta --extern encase=/home/bevy-v0_15_10/target/debug/deps/libencase-2aef43efb1ac2cde.rmeta --extern futures_lite=/home/bevy-v0_15_10/target/debug/deps/libfutures_lite-24084392f2740c7d.rmeta --extern image=/home/bevy-v0_15_10/target/debug/deps/libimage-85916554ad07a057.rmeta --extern ktx2=/home/bevy-v0_15_10/target/debug/deps/libktx2-e54fcf7a088476a9.rmeta --extern naga=/home/bevy-v0_15_10/target/debug/deps/libnaga-97a06b0a35e34f51.rmeta --extern naga_oil=/home/bevy-v0_15_10/target/debug/deps/libnaga_oil-b045d5f4e7484180.rmeta --extern nonmax=/home/bevy-v0_15_10/target/debug/deps/libnonmax-558929e7ac83308a.rmeta --extern offset_allocator=/home/bevy-v0_15_10/target/debug/deps/liboffset_allocator-b855d9d21a4834bf.rmeta --extern serde=/home/bevy-v0_15_10/target/debug/deps/libserde-cb1daf4b77043c9f.rmeta --extern smallvec=/home/bevy-v0_15_10/target/debug/deps/libsmallvec-bdeff86c1ddfc4d2.rmeta --extern wgpu=/home/bevy-v0_15_10/target/debug/deps/libwgpu-91ca241308d34aab.rmeta --cap-lints allow -L native=/home/bevy-v0_15_10/target/debug/build/blake3-69c6c802aad066e9/out -L native=/home/bevy-v0_15_10/target/debug/build/blake3-69c6c802aad066e9/out` (signal: 9, SIGKILL: kill)
warning: build failed, waiting for other jobs to finish...
error: could not compile `bevy_pbr` (lib)

Caused by:
  process didn't exit successfully: `/usr/local/rustup/toolchains/1.83.0-x86_64-unknown-linux-gnu/bin/rustc --crate-name bevy_pbr --edition=2021 /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/bevy_pbr-0.15.0/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=156 --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no -C debuginfo=2 '--warn=clippy::unwrap_or_default' --warn=unused_qualifications --warn=unsafe_op_in_unsafe_fn --deny=unsafe_code --warn=unexpected_cfgs '--warn=clippy::undocumented_unsafe_blocks' '--allow=clippy::type_complexity' '--allow=clippy::too_long_first_doc_paragraph' '--warn=clippy::std_instead_of_core' '--warn=clippy::std_instead_of_alloc' '--warn=clippy::semicolon_if_nothing_returned' '--warn=clippy::ref_as_ptr' '--warn=clippy::redundant_else' '--warn=clippy::redundant_closure_for_method_calls' '--warn=clippy::ptr_cast_constness' '--warn=clippy::ptr_as_ptr' --warn=missing_docs '--warn=clippy::match_same_arms' '--warn=clippy::manual_let_else' '--warn=clippy::doc_markdown' '--warn=clippy::alloc_instead_of_core' --check-cfg 'cfg(docsrs_dep)' -C debug-assertions=on --cfg 'feature="webgl"' --check-cfg 'cfg(docsrs)' --check-cfg 'cfg(feature, values("experimental_pbr_pcss", "ios_simulator", "meshlet", "meshlet_processor", "pbr_anisotropy_texture", "pbr_multi_layer_material_textures", "pbr_transmission_textures", "shader_format_glsl", "trace", "webgl", "webgpu"))' -C metadata=48e7bb2ba9ce0bcd -C extra-filename=-48e7bb2ba9ce0bcd --out-dir /home/bevy-v0_15_10/target/debug/deps -L dependency=/home/bevy-v0_15_10/target/debug/deps --extern bevy_app=/home/bevy-v0_15_10/target/debug/deps/libbevy_app-c41f08641878c131.rmeta --extern bevy_asset=/home/bevy-v0_15_10/target/debug/deps/libbevy_asset-db70c7fbbd812e7f.rmeta --extern bevy_color=/home/bevy-v0_15_10/target/debug/deps/libbevy_color-b63f193e85a44ec9.rmeta --extern bevy_core_pipeline=/home/bevy-v0_15_10/target/debug/deps/libbevy_core_pipeline-46d1861b8b39e3be.rmeta --extern bevy_derive=/home/bevy-v0_15_10/target/debug/deps/libbevy_derive-a36bff48cbb6d803.so --extern bevy_ecs=/home/bevy-v0_15_10/target/debug/deps/libbevy_ecs-761c0185a1c52c72.rmeta --extern bevy_image=/home/bevy-v0_15_10/target/debug/deps/libbevy_image-ca2a43002d3b51e0.rmeta --extern bevy_math=/home/bevy-v0_15_10/target/debug/deps/libbevy_math-b3da14759206059a.rmeta --extern bevy_reflect=/home/bevy-v0_15_10/target/debug/deps/libbevy_reflect-febda5a40b921977.rmeta --extern bevy_render=/home/bevy-v0_15_10/target/debug/deps/libbevy_render-c018e806ddc7dfc5.rmeta --extern bevy_transform=/home/bevy-v0_15_10/target/debug/deps/libbevy_transform-18e0bc892927c0af.rmeta --extern bevy_utils=/home/bevy-v0_15_10/target/debug/deps/libbevy_utils-a3c50cecef0864d2.rmeta --extern bevy_window=/home/bevy-v0_15_10/target/debug/deps/libbevy_window-7d2b86834f475524.rmeta --extern bitflags=/home/bevy-v0_15_10/target/debug/deps/libbitflags-08bd5ef539515700.rmeta --extern bytemuck=/home/bevy-v0_15_10/target/debug/deps/libbytemuck-9ef1f0f72b61bec0.rmeta --extern derive_more=/home/bevy-v0_15_10/target/debug/deps/libderive_more-15fce4e52511f117.rmeta --extern fixedbitset=/home/bevy-v0_15_10/target/debug/deps/libfixedbitset-dc1d487a1f330856.rmeta --extern nonmax=/home/bevy-v0_15_10/target/debug/deps/libnonmax-558929e7ac83308a.rmeta --extern radsort=/home/bevy-v0_15_10/target/debug/deps/libradsort-a54c803b76b1d78a.rmeta --extern smallvec=/home/bevy-v0_15_10/target/debug/deps/libsmallvec-bdeff86c1ddfc4d2.rmeta --extern static_assertions=/home/bevy-v0_15_10/target/debug/deps/libstatic_assertions-6cef7019f3e344fd.rmeta --cap-lints allow -L native=/home/bevy-v0_15_10/target/debug/build/blake3-69c6c802aad066e9/out -L native=/home/bevy-v0_15_10/target/debug/build/blake3-69c6c802aad066e9/out` (signal: 9, SIGKILL: kill)
```

## Checking this with no optimization option
Trying to run this without optimization, i.e., with the following Dockerfile:
```Dockerfile
FROM rust:1.83-bullseye

RUN cargo new /home/bevy-v0_15_10
WORKDIR /home/bevy-v0_15_10

# Create Cargo.toml
RUN echo "[package] \n\
name = 'my_bevy_game' \n\
version = '0.1.0' \n\
edition = '2021' \n\
\n\
[dependencies] \n\
bevy = '0.15.0'" > "Cargo.toml"

RUN apt-get update && apt-get install -y \
        libwayland-dev libxkbcommon-dev \
        g++ pkg-config libx11-dev libasound2-dev libudev-dev libxkbcommon-x11-0
```
and then in the terminal running:
```bash
docker build . -t no-opt-bevy:v0.15.0
docker run -it no-opt-bevy:v0.15.0
# now inside the docker container
cargo build
```

and this builds fine:
```
Finished `dev` profile [unoptimized + debuginfo] target(s) in 2m 09s
```

## The Bug
As described above the build fails for me with optimization, but build well without optimization.
I would like to build my project with optimization ^^'
Would love to hear any idea as to why this happen and how to fix it?
Many thanks

## The Solution
After some trial and error I found out that I should increase the memory capacity available for WSL2 in the ```%userprofile%/.wslconfig```. It was 4GB and I increased it to 8GB
```
[wsl2]
memory=8GB
```
Then, compilation with optimization worked as expected. I monitored the container memory usage in Docker Desktop and it reached around 5.6GB out of the available memory.
I am unsure why this fixes the problem, but it fixes it nonetheless.

## Note
For any future reference, I will note that for me, compilation without optimization takes around 2 minutes, while compilation with optimization takes around 13 minutes. I suppose that any incremental compilation of the project would be faster, but this is still feels like a long time for initial compilation and I wonder if it's worth it.

**Comment by BenjaminBrienen** [comment]
Can you double check that the container has the necessary graphics-related dependencies installed? Check our CI to see what we install there.

**Comment by amam3214** [comment]
Hey, I am quite new to to this so I am unsure what I am looking for in the [bevy CI](https://github.com/bevyengine/bevy_github_ci_template/).
Are you referring to [this line in the ci.yaml?](https://github.com/bevyengine/bevy_github_ci_template/blob/da394f3c96b3821c728648d9b30d7accca27c265/.github/workflows/ci.yaml#L34).
If so, then since my base image is Rust for Debian Bullseye I made sure to and the [needed dependencies](https://github.com/bevyengine/bevy/blob/latest/docs/linux_dependencies.md#ubuntu) as specified in the docs which include these dependencies too. From what I know Ubuntu and Debian are usually similar dependencies-wise so I hope this is not the problem.
And as I mentioned, I don't get any error when compiling without optimization so I don't think it's a dependencies problem. 

I get the same error also when trying to compile on my host machine (WSL Ubuntu) with optimization options, so I tried to build bevy inside a container to try and isolate the problem.
